src/checkers/inference/InferenceTreeAnnotator.java

```diff
@@ -114,13 +117,23 @@ public Void visitClass(final ClassTree classTree, final AnnotatedTypeMirror clas
         InferenceUtil.testArgument(classType instanceof AnnotatedDeclaredType,
                 "Unexpected type for ClassTree ( " + classTree + " ) AnnotatedTypeMirror ( " + classType + " ) ");
 
-        // For anonymous classes, we do not create additional variables, as they
-        // were already handled by the visitNewClass. This would otherwise result
-        // in new variables for an extends clause, which then cannot be inserted.
-        if (!InferenceUtil.isAnonymousClass(classTree)) {
-            this.variableAnnotator.visit(classType, classTree);
+        // Annotated the enclosing type if it exists
+        AnnotatedDeclaredType declType = (AnnotatedDeclaredType) classType;
+        AnnotatedDeclaredType enclosingType = declType.getEnclosingType();
+        TreePath classPath = atypeFactory.getPath(classTree);
+        if (classPath != null) {
+            ClassTree enclosingClass = TreeUtils.enclosingClass(classPath.getParentPath());
+            if (enclosingType != null && enclosingClass != null) {
+                variableAnnotator.visit(enclosingType, enclosingClass);
+            }
         }
 
+        // For anonymous classes, also fully annotate the classTree, which will be
+        // used in the class type validation. Note that this will result in new
+        // variables for extends/implements clauses, so the these variables should
```

Comment:
`for extends/implements clauses`
This sentence is unclear. Where does this happen? None of the calls of `setInsertable` seem to be for extends/implements.